lib/class-wp-rest-pattern-directory-controller.php

```diff
@@ -0,0 +1,209 @@
+		if ( ! current_user_can( 'edit_posts' ) ) {
+			// todo test that contributors can access
+
+			return new WP_Error(
+				'rest_pattern_directory_cannot_view',
+				__( 'Sorry, you are not allowed to browse the local block pattern directory. Please use https://api.wordpress.org/patterns/1.0/ for publicly-available pattern data.', 'gutenberg' ),
+				array( 'status' => rest_authorization_required_code() )
+			);
+		}
+
+		return true;
+	}
+
+	/**
+	 * Search and retrieve block patterns metadata
+	 *
+	 * @since 5.7.0
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 *
+	 * @return WP_Error|WP_REST_Response Response object on success, or WP_Error object on failure.
+	 */
+	public function get_items( $request ) {
+		$action = 'query_patterns';
+		// todo set conditionally ^ based on request
+
+		$api_url = add_query_arg(
+			array(
+				'action'   => $action,
+				'search'   => $request['term'],
+//				'per_page' => $request['per_page'],
+//				'page'     => $request['page'],
+
+				// get_patterns - all
+				// get_patterns &category={id} - get all for category
+				// query_patterns &search=query
+				// set different params for `browse` requests
+			),
+			'http://api.wordpress.org/patterns/1.0/'
+		);
+
+		if ( wp_http_supports( array( 'ssl' ) ) ) {
+			$api_url = set_url_scheme( $api_url, 'https' );
+		}
+
+		// todo cache results locally for 15-30 minutes? or even just 5?
+			// api.w.org has caching, so don't need to worry about it's stability
+			// also don't need to worry about api.w.org return fast
+			// but could worry about the overhead
+
+		$response     = wp_remote_get( $api_url );
+		$raw_patterns = json_decode( wp_remote_retrieve_body( $response ) );
+
+		if ( is_wp_error( $response ) ) {
+			$response->add_data( array( 'status' => 500 ) );
+
+			return $response;
+		}
+
+		$prepared_patterns = array();
+
+		if ( $raw_patterns ) {
+			foreach ( $raw_patterns as $pattern ) {
+				$prepared_patterns[] = $this->prepare_item_for_response( $pattern, $action );
+			}
+		}
+
+		return rest_ensure_response( $prepared_patterns );
+	}
+
+	/**
+	 * Parse pattern metadata for a pattern, and prepare it for an API response.
+	 *
+	 * @since 5.7.0
+	 *
+	 * @param array  $raw_pattern A pattern from api.wordpress.org, before any changes.
+	 * @param string $action      The type of request that resulted in the raw patterns. Either 'get_patterns' or 'query_patterns'.
+	 *
+	 * @return array
+	 */
+	public function prepare_item_for_response( $raw_pattern, $action ) {
+		switch ( $action ) {
+			case 'get_patterns':
+				// todo filter by different fields than search
+				// need to result in the exact same schema though?
+				break;
+
+			case 'query_patterns':
+				$prepared_pattern = (array) $raw_pattern->{'_embedded'}->self[0];
+				$prepared_pattern = array_intersect_key( $prepared_pattern, array_flip( self::SAFELISTED_FIELDS ) );
```

Comment:
Instead of pulling from a separate constant, I'd look at the schema's properties.

Reply:
👍🏻

Reply:
Fixed in 6226e18